Wire up the history tracking delegate for GeckoView

linacambridge opened the pull request:
I tested this with my local GeckoView build, the sample browser, and @grigoryk's in-memory history storage, and it works! 🎉 There's also a commit that points the Nightly x86 target to m-c instead, for folks to try out if they'd like. 😄

pocmo commented:
GV bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1494713

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
+                var exception: Throwable? = null
+                val visits: List<Boolean>? = try {
+                    settings.historyTrackingDelegate?.getVisited(urls.toList(), privateMode)?.await()
+                } catch (t: Throwable) {
```
Generally you watch to catch as narrow set of exceptions as possible. `Throwable` covers unchecked exceptions (subclasses of `RuntimeException` and `Error`, such as `NullPointerException`, `ThreadDeath`, etc), which is not what we want here. If there's nothing better, use `Exception` instead.

linacambridge replied:
Hmm, I'm not sure what all we'd like to catch here; IIUC, `getVisited` can throw anything. On the one hand, I don't think we want to crash if `getVisited` throws; on the other, I don't see other handlers doing the same, and, given that Gecko's [`OnError` handlers do nothing](https://searchfox.org/mozilla-central/rev/0859e6b10fb901875c80de8f8fc33cbb77b2505e/mobile/android/components/geckoview/GeckoViewHistory.cpp#363-367,565-569), maybe it's best to remove this entirely.

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
+                    settings.historyTrackingDelegate?.getVisited(urls.toList(), privateMode)?.await()
+                } catch (t: Throwable) {
+                    exception = t
+                    null
```
Remove this?

linacambridge replied:
I'm assigning to `val visits`, so I think I still need the `null`, but it's moot if I remove the catches.

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
+            return result
+        }
+
+        @Suppress("TooGenericExceptionCaught")
```
Hopefully this will go away if you don't catch `throwable`.

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
+            scope.launch {
+                var exception: Throwable? = null
+                try {
+                    settings.historyTrackingDelegate?.onVisited(url, isReload, privateMode)
```
We shouldn't even call into `historyTrackingDelegate` if we're in a `privateMode`.

linacambridge replied:
In that case, should we also remove the `privateMode` argument from `historyTrackingDelegate.{onVisited, getVisited}`?

grigoryk replied:
On one hand, `privateMode` is useful on a conceptual level in the delegate as it allows to easily segment history into two buckets, and let the storage layer do something sensible with those buckets. On another hand, we don't actually follow through on this, and it's not clear that we'll want to. I'll file a follow-up to consider cleaning this up.

grigoryk replied:
Ah! I see that you killed that argument. That's fine!

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -57,6 +61,13 @@ class GeckoEngineSession(
 
     private var initialLoad = true
 
+    private val scope by lazy { CoroutineScope(Dispatchers.IO) }
```
nit: Can you change other users of the history delegate to use this scope as well?

linacambridge replied:
Do you mean I should make this public and have `PlacesHistoryStorage` use this delegate? How would that work if `PlacesHistoryStorage` doesn't access the session; should I pass the scope as an argument to `GeckoEngineSession` and `PlacesHistoryStorage` instead?

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
+
+            val isReload = lastVisitedURL?.let { it == url } ?: false
+            val result = GeckoResult<Boolean>()
+            scope.launch {
```
This should not do any work if: 
- we're in privateMode
- there's no configured `historyTrackingDelegate`

In those two cases, you'd want to resolve GeckoResult right away without doing anything with scopes.

grigoryk commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
+        }
+
+        @Suppress("TooGenericExceptionCaught")
+        override fun getVisited(
```
Same as in `onVisited`, resolve this right away if we don't have a configured delegate or if we're in a private mode.

linacambridge commented:
@grigoryk Whew! I think this is ready for re-review. Changes since last time:

* Removed the `privateMode` argument from `HistoryTrackingDelegate`. Instead, we'll skip the call to the delegate entirely.
* Bail early instead of launching a coroutine if a delegate isn't set.
* Tests actually work! 🎉
* Made both overloads of `getVisited` `suspend` functions.

linacambridge commented:
Looks like the `browser-awesomebar` failure is #1300; I didn't see it fail on previous runs.

grigoryk commented:
This landed!
